net/instaweb/rewriter/csp_test.cc

```diff
@@ -65,31 +85,119 @@ TEST(CspParseSourceTest, NonQuoted) {
+            CspSourceExpression::kHostSource,
+            CspSourceExpression::UrlData("http", "www.example.com", "",
+                                         "/dir")),
       CspSourceExpression::Parse("http://www.example.com/dir"));
 
   EXPECT_EQ(
-      CspSourceExpression(CspSourceExpression::kHostSource,
-                          "http://www.example.com/dir/file.js"),
+      CspSourceExpression(
+            CspSourceExpression::kHostSource,
+            CspSourceExpression::UrlData("http", "www.example.com", "",
+                                         "/dir/file.js")),
       CspSourceExpression::Parse("http://www.example.com/dir/file.js"));
 
   EXPECT_EQ(
-      CspSourceExpression(CspSourceExpression::kHostSource, "*"),
+      CspSourceExpression(
+            CspSourceExpression::kHostSource,
+            CspSourceExpression::UrlData("", "www.example.com", "",
+                                         "/dir/file.js")),
+      CspSourceExpression::Parse("www.example.com/dir/file.js"));
+
+  EXPECT_EQ(
+      CspSourceExpression(CspSourceExpression::kHostSource,
+                          CspSourceExpression::UrlData("", "*", "", "")),
       CspSourceExpression::Parse("*"));
+
+  EXPECT_EQ(
+      CspSourceExpression(CspSourceExpression::kUnknown),
+      CspSourceExpression::Parse("http:!/example.com"));
+
+  EXPECT_EQ(
+      CspSourceExpression(CspSourceExpression::kUnknown),
+      CspSourceExpression::Parse("http://"));
+
+  EXPECT_EQ(
+      CspSourceExpression(CspSourceExpression::kUnknown),
+      CspSourceExpression::Parse("http:/"));
+
+  EXPECT_EQ(
+      CspSourceExpression(CspSourceExpression::kUnknown),
+      CspSourceExpression::Parse("http:/example.com"));
+
+  EXPECT_EQ(
+      CspSourceExpression(CspSourceExpression::kUnknown),
+      CspSourceExpression::Parse("?example.com/dir/file.js"));
+
+  EXPECT_EQ(
+      CspSourceExpression(CspSourceExpression::kUnknown),
+      CspSourceExpression::Parse("http:///dir/file.js"));
+
+  EXPECT_EQ(
+      CspSourceExpression(
+          CspSourceExpression::kHostSource,
+          CspSourceExpression::UrlData("https", "*", "*", "/foo.js")),
+      CspSourceExpression::Parse("https://*:*/foo.js"));
+
+  // Test for no port after :. Note that this needs an explicit scheme, since
+  // www.example.com: would be a valid scheme-source!
+  EXPECT_EQ(
+      CspSourceExpression(CspSourceExpression::kUnknown),
+      CspSourceExpression::Parse("http://www.example.com:"));
+
+  EXPECT_EQ(
+      CspSourceExpression(CspSourceExpression::kUnknown),
+      CspSourceExpression::Parse("www.example.com:/foo"));
+
+  EXPECT_EQ(
+      CspSourceExpression(
+          CspSourceExpression::kHostSource,
+          CspSourceExpression::UrlData("https", "*", "443", "/foo.js")),
+      CspSourceExpression::Parse("https://*:443/foo.js"));
+
+  EXPECT_EQ(
+      CspSourceExpression(CspSourceExpression::kUnknown),
+      CspSourceExpression::Parse("https://*:443?foo.js"));
+}
+
```

Comment:
`CspSourceExpression::ParseQuoted("\"\"")` -> should have been `CspSourceExpression::ParseQuoted("''")`
But either way, I noticed there's already a test that inputs only whitespace, and this one wouldn't add value on top of that, so please ignore